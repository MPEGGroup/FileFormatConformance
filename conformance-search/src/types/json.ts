export type Coverage = {
    not_found: {
        count: number;
        percentage: number;
        boxes: string[];
        missing_extensions: string[];
        paths: {
            [path: string]: string[];
        };
    };
    lists: {
        boxes: {
            covered: string[];
            not_covered: string[];
            under_consideration: string[];
        };
        features: {
            covered: string[];
            not_covered: string[];
            under_consideration: string[];
        };
    };
    path_file_map: {
        count: number;
        non_empty: number;
        overall_coverage: number;
        published: {
            count: number;
            coverage: number;
        };
        under_consideration: {
            count: number;
            coverage: number;
        };
    };
    feature_file_map: {
        count: number;
        published_features: number;
        under_consideration_features: number;
    };
    files: {
        count: number;
        published_count: number;
        under_consideration_count: number;
    };
    boxes: {
        count: number;
    };
    spec_info: {
        ISO: string;
        name: string;
        version: number;
        date: number;
        mp4ra_reference: string[];
        link: string;
        amendmends?: string[];
        corrigenda?: string[];
    }[];
};

export type FileMetadata = {
    contributor: string;
    description: string;
    license: string;
    md5: string;
    rel_filepath: string;
    abs_filepath: string;
    version: number;
    published: boolean;
    associated_files: string[];
    mdms_number: string;
    features: string[];
    conforms_to: {
        ISO: string;
        version: number;
        amendmends?: string[];
        corrigenda?: string[];
    }[];
    notes:
        | string
        | {
              [key: string]: unknown;
          };
};

export type FeatureFileMap = {
    key: string;
    description: string;
    spec: string;
    files: string[];
};

export type Files = {
    not_found: string[];
    path_file_map: {
        [path: string]: {
            [variant: string]: string[];
        };
    };
    feature_file_map: {
        [feature: string]: FeatureFileMap;
    };
    file_metadata: {
        [file: string]: FileMetadata;
    };
    boxes: {
        [box: string]: string[][];
    };
};

export type Dictionary = {
    fourccs: string[];
    types: string[];
    specs: string[];
    registered_types: string[];
};

export type Hierarchy = {
    [key: string]: Hierarchy;
};
