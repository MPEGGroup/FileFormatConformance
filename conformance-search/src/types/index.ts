import Fuse from "fuse.js";
import { FileMetadata } from "./json";

export type Box = {
    fourcc: string;
    spec: string;
    codec: string;
    description: string;
    type: string;
    versions: number[];
    flags: (
        | {
              name: string;
              value: string;
              description: string;
          }
        | {
              wildcard: string;
          }
    )[];
    containers: {
        fourcc: string;
        type: string;
    }[];
    syntax: string;
};

export type Feature = {
    name: string;
    description: string;
    spec: string;
};

export type Filter = {
    type: string;
    value: string;
};

export type FilterConverted = {
    [key: string]: string;
};

export type SearchResult<T> = {
    item: T;
    score: number;
    actualScore: number;
    refIndex: number;
};

export type SearchResultRefined<T> = SearchResult<T> & {
    count: number;
    type: string;
    exclude: boolean;
    refinements?: {
        variant: {
            versions: {
                value: number[];
            };
            flags: {
                value: string[];
                exact: boolean;
            };
        };
    };
};

export type FuseSearchWithScore<T> = Fuse.FuseResult<T> & {
    score: number;
    actualScore: number;
};

export type FileSearchResult = {
    item: FileMetadata;
    matches: {
        type: string;
        value: string;
    }[];
};

export type CoverageResults = {
    boxes: {
        covered: Fuse.FuseResult<string>[];
        not_covered: Fuse.FuseResult<string>[];
        under_consideration: Fuse.FuseResult<string>[];
        [key: string]: Fuse.FuseResult<string>[];
    };
    features: {
        covered: Fuse.FuseResult<string>[];
        not_covered: Fuse.FuseResult<string>[];
        under_consideration: Fuse.FuseResult<string>[];
        [key: string]: Fuse.FuseResult<string>[];
    };
};
