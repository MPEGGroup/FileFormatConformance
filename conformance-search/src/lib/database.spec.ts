import { Box, SearchResultRefined } from "@/types";
import Database from "./database";
import Search from "./search";

describe("Database", () => {
    let database: Database;
    let search: Search;
    beforeAll(async () => {
        database = await Database.getInstance();
        search = await Search.getInstance();
    });

    function getBox(fourcc: string): SearchResultRefined<Box> {
        const box = database.boxes.find((obox: Box) => obox.fourcc === fourcc);
        if (!box) throw new Error(`Box ${fourcc} not found in database`);
        return {
            item: box,
            refIndex: -1,
            actualScore: 0,
            score: 0,
            count: 0,
            exclude: false,
            type: "box"
        };
    }

    describe("files", () => {
        beforeEach(async () => {
            await database.updateDatabase([]);
        });

        it("should return 3 files for stz2", () => {
            const results = database.getMatchingFilesByBox([getBox("stz2")]);

            expect(results.length).toEqual(3);
            expect(results.map((r) => r.item.abs_filepath)).toMatchSnapshot();
        });

        it("should return 2 files for meta under trak", async () => {
            await database.updateDatabase([{ container: "$trak" }]);
            const results = database.getMatchingFilesByBox([getBox("meta")]);

            expect(results.length).toEqual(2);
            expect(results.map((r) => r.item.abs_filepath)).toMatchSnapshot();
        });

        it("should return 12 files for idat under =file.meta", async () => {
            await database.updateDatabase([{ container: "=$file.$meta" }]);
            const results = database.getMatchingFilesByBox([getBox("idat")]);

            expect(results.length).toEqual(12);
            expect(results.map((r) => r.item.abs_filepath)).toMatchSnapshot();
        });

        it("should return 14 files for idat under file.meta", async () => {
            await database.updateDatabase([{ container: "$file.$meta" }]);
            const results = database.getMatchingFilesByBox([getBox("idat")]);

            expect(results.length).toEqual(14);
            expect(results.map((r) => r.item.abs_filepath)).toMatchSnapshot();
        });

        it("should return 5 files for hdlr under moov.FullBox", async () => {
            await database.updateDatabase([{ container: "$moov.#FullBox" }]);
            const results = database.getMatchingFilesByBox([getBox("hdlr")]);

            expect(results.length).toEqual(5);
            expect(results.map((r) => r.item.abs_filepath)).toMatchSnapshot();
        });

        it("should return 5 files for hdlr under moov.meta", async () => {
            await database.updateDatabase([{ container: "$moov.$meta" }]);
            const results = database.getMatchingFilesByBox([getBox("hdlr")]);

            expect(results.length).toEqual(5);
            expect(results.map((r) => r.item.abs_filepath)).toMatchSnapshot();
        });

        it("should return 3 files for hdlr under =moov.meta", async () => {
            await database.updateDatabase([{ container: "=$moov.$meta" }]);
            const results = database.getMatchingFilesByBox([getBox("hdlr")]);

            expect(results.length).toEqual(3);
            expect(results.map((r) => r.item.abs_filepath)).toMatchSnapshot();
        });

        it("should return 2 files for hdlr under =moov.trak.meta", async () => {
            await database.updateDatabase([{ container: "=$moov.$trak.$meta" }]);
            const results = database.getMatchingFilesByBox([getBox("hdlr")]);

            expect(results.length).toEqual(2);
            expect(results.map((r) => r.item.abs_filepath)).toMatchSnapshot();
        });

        it("should return more than 0 files after removing a container filter", async () => {
            await search.search("meta", [{ type: "container", value: "$trak" }]);
            const { boxes } = await search.search("mvex", []);
            const wrappedBoxes = boxes.map((box) => getBox(box.item.fourcc));

            const files = database.getMatchingFilesByBox(wrappedBoxes);

            expect(files.length).toBeGreaterThan(0);
        });
    });

    describe("dictionary", () => {
        it("should return a correct snapshot for =$trak", async () => {
            await database.updateDatabase([{ container: "=$trak" }]);
            const sorted = {
                fourccs: database.dictionary.fourccs.sort(),
                types: database.dictionary.types.sort()
            };
            expect(sorted).toMatchSnapshot();
        });

        it("should return a correct snapshot for =$moov", async () => {
            await database.updateDatabase([{ container: "=$moov" }]);
            const sorted = {
                fourccs: database.dictionary.fourccs.sort(),
                types: database.dictionary.types.sort()
            };
            expect(sorted).toMatchSnapshot();
        });
    });
});
