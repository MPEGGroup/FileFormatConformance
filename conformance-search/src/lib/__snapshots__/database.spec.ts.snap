// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Database > dictionary > should return a correct snapshot for =$moov 1`] = `
{
  "fourccs": [
    "free",
    "meta",
    "mvex",
    "mvhd",
    "skip",
    "trak",
    "udta",
    "uuid",
  ],
  "types": [
    "Box",
    "FullBox",
  ],
}
`;

exports[`Database > dictionary > should return a correct snapshot for =$trak 1`] = `
{
  "fourccs": [
    "edts",
    "free",
    "mdia",
    "meta",
    "skip",
    "tkhd",
    "tref",
    "trgr",
    "ttyp",
    "udta",
    "uuid",
  ],
  "types": [
    "Box",
    "FullBox",
    "GeneralTypeBox",
  ],
}
`;

exports[`Database > files > should return 2 files for hdlr under =moov.trak.meta 1`] = `
[
  "published/isobmff/12_metas_v2.mp4",
  "published/maf/vsaf/1.mp4",
]
`;

exports[`Database > files > should return 2 files for meta under trak 1`] = `
[
  "published/isobmff/12_metas_v2.mp4",
  "published/maf/vsaf/1.mp4",
]
`;

exports[`Database > files > should return 3 files for hdlr under =moov.meta 1`] = `
[
  "published/isobmff/12_metas_v2.mp4",
  "published/uvvu/Solekai002_1280_23_1x1_v7clear.uvvu",
  "published/uvvu/Solekai007_1920_29_1x1_v7clear.uvu",
]
`;

exports[`Database > files > should return 3 files for meta under =moov 1`] = `
[
  "published/isobmff/12_metas_v2.mp4",
  "published/uvvu/Solekai002_1280_23_1x1_v7clear.uvvu",
  "published/uvvu/Solekai007_1920_29_1x1_v7clear.uvu",
]
`;

exports[`Database > files > should return 3 files for stz2 1`] = `
[
  "published/isobmff/02_dref_edts_img.mp4",
  "published/isobmff/04_bifs_video.mp4",
  "published/isobmff/a7-tone-oddities.mp4",
]
`;

exports[`Database > files > should return 5 files for hdlr under moov.FullBox 1`] = `
[
  "published/isobmff/12_metas_v2.mp4",
  "published/uvvu/Solekai002_1280_23_1x1_v7clear.uvvu",
  "published/uvvu/Solekai007_1920_29_1x1_v7clear.uvu",
  "published/maf/vsaf/1.mp4",
  "published/isobmff/restricted.mp4",
]
`;

exports[`Database > files > should return 5 files for hdlr under moov.meta 1`] = `
[
  "published/isobmff/12_metas_v2.mp4",
  "published/uvvu/Solekai002_1280_23_1x1_v7clear.uvvu",
  "published/uvvu/Solekai007_1920_29_1x1_v7clear.uvu",
  "published/maf/vsaf/1.mp4",
  "published/isobmff/restricted.mp4",
]
`;

exports[`Database > files > should return 12 files for idat under =file.meta 1`] = `
[
  "published/heif/C007.heic",
  "published/heif/C015.heic",
  "published/heif/C016.heic",
  "published/heif/C017.heic",
  "published/heif/C018.heic",
  "published/heif/C019.heic",
  "published/heif/C020.heic",
  "published/heif/C021.heic",
  "published/heif/C022.heic",
  "published/heif/C023.heic",
  "published/heif/C024.heic",
  "published/heif/C025.heic",
]
`;

exports[`Database > files > should return 14 files for idat under file.meta 1`] = `
[
  "published/heif/C007.heic",
  "published/heif/C015.heic",
  "published/heif/C016.heic",
  "published/heif/C017.heic",
  "published/heif/C018.heic",
  "published/heif/C019.heic",
  "published/heif/C020.heic",
  "published/heif/C021.heic",
  "published/heif/C022.heic",
  "published/heif/C023.heic",
  "published/heif/C024.heic",
  "published/heif/C025.heic",
  "published/uvvu/Solekai002_1280_23_1x1_v7clear.uvvu",
  "published/uvvu/Solekai007_1920_29_1x1_v7clear.uvu",
]
`;
